### PR TITLE
[batch] Use SQL database in server.py

### PR DESCRIPTION
**Pull request by jigold**
Depends on #5791 and #5809

### CODE REVIEW

**Comment by jigold** [comment]
I have some small things to work out with this PR. I'll let you know when it's ready to be reviewed.

**Comment by jigold** [comment]
@danking Ready for review.

File: batch/batch/database.py [hunk 1]
```diff
@@ -13,9 +13,10 @@ def run_synchronous(coro):
 
 @asyncinit
 class Database:
-    @staticmethod
-    def create_synchronous(config_file):
-        db = run_synchronous(Database(config_file))
+    @classmethod
+    def create_synchronous(cls, config_file):
+        db = object.__new__(cls)
+        run_synchronous(cls.__init__(db, config_file))
```
Comment (danking) [inline batch/batch/database.py]: can you comment on why this is necessary?

File: batch/batch/server/database.py [hunk 2]
```diff
@@ -64,12 +57,26 @@ class JobsTable(Table):
...
-        parent_ids = json.loads(parent_ids['parent_ids'])
-        parent_records = await self.get_records(parent_ids)
-        incomplete_parents = [pr['id'] for pr in parent_records.result()
-                              if pr['state'] == 'Created']
-        return incomplete_parents
+        async with self._db.pool.acquire() as conn:
+            async with conn.cursor() as cursor:
+                sql = f"""SELECT `id` FROM `{self.name}` WHERE `id` IN
+                          (SELECT `parent_id` FROM `{self._db.jobs_parents.name}` WHERE `job_id` = %s)
+                          AND `state` = %s""".replace('\n', ' ')
```
Comment (danking) [inline batch/batch/server/database.py]: It's a bit unusual to see query as a subquery rather than a join, but I suppose it doesn't much matter.
```
SELECT {self._db.jobs_parents.name}.parent_id
FROM {self.name}.id
INNER JOIN {self._db.jobs_parents.name} ON {self.name}.id = {self._db.jobs_parents.name}.id
WHERE {self.name}.state = %s AND {self._db.jobs_parents.name}.job_id = %s
```

File: batch/batch/server/database.py [hunk 3]
```diff
@@ -64,12 +57,26 @@ class JobsTable(Table):
...
+                return [record['id'] for record in result]
+
+    async def get_record_by_pod(self, pod):
+        records = await super(JobsTable, self).get_record({'pod_name': pod})
+        if len(records) == 0:  # pylint: disable=R1705
+            return None
+        elif len(records) == 1:
+            return records[0]
+        else:
+            raise Exception("'jobs' table error. Cannot have same pod in more than one record")
```
Comment (danking) [inline batch/batch/server/database.py]: include the first few records in the message for debugging help

File: batch/batch/server/database.py [hunk 4]
```diff
@@ -64,12 +57,26 @@ class JobsTable(Table):
...
+            async with conn.cursor() as cursor:
+                sql = f"""SELECT `id` FROM `{self.name}` WHERE `id` IN
+                          (SELECT `parent_id` FROM `{self._db.jobs_parents.name}` WHERE `job_id` = %s)
+                          AND `state` = %s""".replace('\n', ' ')
+                await cursor.execute(sql, (id, 'Created'))
+                result = await cursor.fetchall()
+                return [record['id'] for record in result]
+
+    async def get_record_by_pod(self, pod):
+        records = await super(JobsTable, self).get_record({'pod_name': pod})
```
Comment (danking) [inline batch/batch/server/database.py]: `self.get_records_where({'pod_name': pod})`, yeah?

File: batch/batch/server/database.py [hunk 5]
```diff
@@ -80,6 +87,9 @@ class JobsParentsTable(Table):
 
         await super().__init__(db, name, schema, keys)
 
+    async def new_record(self, **items):
```
Comment (danking) [inline batch/batch/server/database.py]: why are these necessary?  does this class not just inherit the operations?

File: batch/test/test_batch.py [hunk 6]
```diff
@@ -134,7 +134,15 @@ def test_deleted_job_log(self):
...
         j.delete()
-        self.assertEqual(self.batch._get_job_log(id), {'main': 'test\n'})
+
+        try:
+            self.batch._get_job_log(id)
+        except requests.HTTPError as e:
+            if e.response.status_code == 404:
+                pass
+            else:
+                raise
```
Comment (danking) [inline batch/test/test_batch.py]: `self.assertTrue(False, f"batch should not have deleted log {e}")` is preferable to indicate test failure

File: batch/batch/database.py (hunk 1 above)
Comment (jigold) [reply batch/batch/database.py]: I think this is clearer than `run_synchronous(Database(config_file))`.

**Comment by jigold** [comment]
@danking  I'm going to close for now until #5866 goes in.
